File: Sources/ModelRuntime/Model/ModelManifest.swift
public struct ModelManifest: Codable, Equatable, Sendable {
    public var schemaVersion: Int
    public var model: ModelInfo
    public var runtime: RuntimeInfo
    public var architecture: ArchitectureInfo
    public var deviceProfiles: [String: DeviceProfileConfiguration]
    public var contextVariants: [Int]
    public var asset: AssetInfo
    public var quantization: QuantizationInfo
    public var fallbackPolicy: FallbackPolicy

    public var validationErrors: [String] {
        var errors: [String] = []

        if model.id != ModelManifestContract.expectedModelId {
            errors.append("model.id must be \(ModelManifestContract.expectedModelId)")
        }

        if runtime.type != ModelManifestContract.expectedRuntime {
            errors.append("runtime.type must be \(ModelManifestContract.expectedRuntime)")
        }

        if !runtime.entrypoints.contains("prefill") || !runtime.entrypoints.contains("decode") {
            errors.append("runtime.entrypoints must include prefill and decode")
        }

        if !ModelManifestContract.supportedKVCacheModes.contains(runtime.kvCacheMode) {
            errors.append("runtime.kvCacheMode must be stateful-preferred, slot-ring, or contiguous-sliding")
        }

        validateRuntimeGraphSchema(into: &errors)

        if architecture.layers != ModelManifestContract.layers {
            errors.append("architecture.layers must be \(ModelManifestContract.layers)")
        }

        if architecture.hiddenSize != ModelManifestContract.hiddenSize {
            errors.append("architecture.hiddenSize must be \(ModelManifestContract.hiddenSize)")
        }

        if architecture.queryHeads != ModelManifestContract.queryHeads {
            errors.append("architecture.queryHeads must be \(ModelManifestContract.queryHeads)")
        }

        if architecture.kvHeads != ModelManifestContract.kvHeads {
            errors.append("architecture.kvHeads must be \(ModelManifestContract.kvHeads)")
        }

        if !architecture.tokenizer.preserved {
            errors.append("tokenizer must be preserved")
        }

        if !architecture.tokenizer.vocabularyPreserved {
            errors.append("vocabulary must be preserved")
        }

        for variant in contextVariants where !ModelManifestContract.supportedContextVariants.contains(variant) {
            errors.append("unsupported context variant \(variant)")
        }

        if asset.storage == "app-bundle" {
            errors.append("asset.storage must not be app-bundle")
        }

        validateStatefulSharedArtifacts(into: &errors)

        if let variants = asset.variants {
            for profile in DeviceProfile.allCases {
                guard let defaultContextVariant = deviceProfiles[profile.rawValue]?.defaultContextVariant else {
                    continue
                }
                guard let variant = variants[String(defaultContextVariant)] else {
                    errors.append("asset.variants.\(defaultContextVariant) must be present for \(profile.rawValue)")
                    continue
                }

                if variant.deviceProfile != profile.rawValue {
                    errors.append("asset.variants.\(defaultContextVariant).deviceProfile must be \(profile.rawValue)")
                }
            }
        }

        if quantization.strategy != "mixed-precision-fidelity-first" {
            errors.append("quantization.strategy must be mixed-precision-fidelity-first")
        }

        if quantization.kvCache != "fp16" && quantization.kvCache != "int8" {
            errors.append("quantization.kvCache must be fp16 or int8")
        }

        if quantization.structuralReduction {
            errors.append("structuralReduction must be false")
        }

        return errors
    }

    private func validateRuntimeGraphSchema(into errors: inout [String]) {
        let graphSchema = runtime.graphSchema

        if !ModelManifestContract.supportedGraphInterfaces.contains(graphSchema.interface) {
            errors.append("runtime.graphSchema.interface must be logits-layered-kv, stateful-kv, or stateful-step-kv")
        }

        if graphSchema.layerCount != ModelManifestContract.layers {
            errors.append("runtime.graphSchema.layerCount must be \(ModelManifestContract.layers)")
        }

        if graphSchema.kvHeads != ModelManifestContract.kvHeads {
            errors.append("runtime.graphSchema.kvHeads must be \(ModelManifestContract.kvHeads)")
        }

        if graphSchema.headDimension != ModelManifestContract.headDimension {
            errors.append("runtime.graphSchema.headDimension must be \(ModelManifestContract.headDimension)")
        }

        if graphSchema.prefill.inputIDs != ModelManifestContract.prefillInputIDs {
            errors.append("runtime.graphSchema.prefill.inputIDs must be \(ModelManifestContract.prefillInputIDs)")
        }

        if graphSchema.prefill.positionIDs != ModelManifestContract.prefillPositionIDs {
            errors.append("runtime.graphSchema.prefill.positionIDs must be \(ModelManifestContract.prefillPositionIDs)")
        }

        if graphSchema.prefill.causalMask != ModelManifestContract.causalMask {
            errors.append("runtime.graphSchema.prefill.causalMask must be \(ModelManifestContract.causalMask)")
        }

        if graphSchema.prefill.logits != ModelManifestContract.logits {
            errors.append("runtime.graphSchema.prefill.logits must be \(ModelManifestContract.logits)")
        }

        if graphSchema.prefill.keyPrefix != ModelManifestContract.prefillKeyPrefix {
            errors.append("runtime.graphSchema.prefill.keyPrefix must be \(ModelManifestContract.prefillKeyPrefix)")
        }

        if graphSchema.prefill.valuePrefix != ModelManifestContract.prefillValuePrefix {
            errors.append("runtime.graphSchema.prefill.valuePrefix must be \(ModelManifestContract.prefillValuePrefix)")
        }

        let expectedDecodeTokenID = ModelManifestContract.expectedDecodeTokenID(for: graphSchema.interface)
        if graphSchema.decode.tokenID != expectedDecodeTokenID {
            errors.append("runtime.graphSchema.decode.tokenID must be \(expectedDecodeTokenID)")
        }

        let expectedDecodePositionID = ModelManifestContract.expectedDecodePositionID(for: graphSchema.interface)
        if graphSchema.decode.positionID != expectedDecodePositionID {
            errors.append("runtime.graphSchema.decode.positionID must be \(expectedDecodePositionID)")
        }

        if graphSchema.decode.causalMask != ModelManifestContract.causalMask {
            errors.append("runtime.graphSchema.decode.causalMask must be \(ModelManifestContract.causalMask)")
        }

        if graphSchema.decode.logits != ModelManifestContract.logits {
            errors.append("runtime.graphSchema.decode.logits must be \(ModelManifestContract.logits)")
        }

        if graphSchema.decode.pastKeyPrefix != ModelManifestContract.decodePastKeyPrefix {
            errors.append("runtime.graphSchema.decode.pastKeyPrefix must be \(ModelManifestContract.decodePastKeyPrefix)")
        }

        if graphSchema.decode.pastValuePrefix != ModelManifestContract.decodePastValuePrefix {
            errors.append("runtime.graphSchema.decode.pastValuePrefix must be \(ModelManifestContract.decodePastValuePrefix)")
        }

        if graphSchema.decode.newKeyPrefix != ModelManifestContract.decodeNewKeyPrefix {
            errors.append("runtime.graphSchema.decode.newKeyPrefix must be \(ModelManifestContract.decodeNewKeyPrefix)")
        }

        if graphSchema.decode.newValuePrefix != ModelManifestContract.decodeNewValuePrefix {
            errors.append("runtime.graphSchema.decode.newValuePrefix must be \(ModelManifestContract.decodeNewValuePrefix)")
        }
    }

    private func validateStatefulSharedArtifacts(into errors: inout [String]) {
        guard ModelManifestContract.statefulGraphInterfaces.contains(runtime.graphSchema.interface) else {
            return
        }

        if asset.prefillPath != asset.decodePath {
            errors.append("stateful Core ML graphs must use the same artifact path for prefill and decode")
        }

        for (contextVariant, variant) in asset.variants ?? [:] where variant.prefillPath != variant.decodePath {
            errors.append("asset.variants.\(contextVariant) must use the same artifact path for prefill and decode for stateful Core ML graphs")
        }
    }

    public func modelArtifact(
        for deviceProfile: DeviceProfile,
        requestedContextTokens: Int?
    ) throws -> SelectedModelArtifact {
        guard let profile = deviceProfiles[deviceProfile.rawValue] else {
            throw InferenceRuntimeError.invalidInput(message: "Unsupported device profile \(deviceProfile.rawValue).")
        }

        let requested = requestedContextTokens ?? profile.defaultContextVariant
        let sortedContextVariants = contextVariants.sorted()
        guard let fallbackContext = sortedContextVariants.first else {
            throw InferenceRuntimeError.invalidInput(message: "Manifest has no context variants.")
        }
        let selectedContext = sortedContextVariants
            .filter { $0 <= requested }
            .last ?? fallbackContext

        if let variant = asset.variants?[String(selectedContext)] {
            return SelectedModelArtifact(
                contextVariant: selectedContext,
                deviceProfile: variant.deviceProfile,
                prefillPath: variant.prefillPath,
                decodePath: variant.decodePath,
                tokenizerPath: variant.tokenizerPath ?? asset.tokenizerPath,
                sha256: variant.sha256,
                prefillSHA256: variant.prefillSHA256,
                decodeSHA256: variant.decodeSHA256,
                tokenizerSHA256: variant.tokenizerSHA256 ?? asset.tokenizerSHA256
            )
        }

        return SelectedModelArtifact(
            contextVariant: selectedContext,
            deviceProfile: deviceProfile.rawValue,
            prefillPath: asset.prefillPath,
            decodePath: asset.decodePath,
            tokenizerPath: asset.tokenizerPath,
            sha256: asset.sha256,
            prefillSHA256: asset.prefillSHA256,
            decodeSHA256: asset.decodeSHA256,
            tokenizerSHA256: asset.tokenizerSHA256
        )
    }
}

public struct ModelInfo: Codable, Equatable, Sendable {
    public var id: String
    public var revision: String
    public var parameterCount: Int
}

public struct RuntimeInfo: Codable, Equatable, Sendable {
    public var type: String
    public var entrypoints: [String]
    public var kvCacheMode: String
    public var graphSchema: RuntimeGraphSchema

    public func kvCacheRouteDecision(
        capabilities: CoreMLRuntimeCapabilities
    ) -> CoreMLKVCacheRouteDecision {
        CoreMLKVCacheRoutePlanner.selectRoute(
            kvCacheMode: kvCacheMode,
            graphInterface: graphSchema.interface,
            capabilities: capabilities
        )
    }

    public var kvCacheUpdateStrategy: KVCacheUpdateStrategy {
        switch kvCacheMode {
        case "contiguous-sliding":
            return .contiguousSliding
        case "stateful-preferred", "slot-ring":
            return .slotRing
        default:
            return .slotRing
        }
    }
}

public struct RuntimeGraphSchema: Codable, Equatable, Sendable {
    public var interface: String
    public var layerCount: Int
    public var kvHeads: Int
    public var headDimension: Int
    public var prefill: PrefillGraphSchema
    public var decode: DecodeGraphSchema
}

public struct PrefillGraphSchema: Codable, Equatable, Sendable {
    public var inputIDs: String
    public var positionIDs: String
    public var causalMask: String
    public var logits: String
    public var keyPrefix: String
    public var valuePrefix: String
}

public struct DecodeGraphSchema: Codable, Equatable, Sendable {
    public var tokenID: String
    public var positionID: String
    public var causalMask: String
    public var logits: String
    public var pastKeyPrefix: String
    public var pastValuePrefix: String
    public var newKeyPrefix: String
    public var newValuePrefix: String
}

public struct ArchitectureInfo: Codable, Equatable, Sendable {
    public var type: String
    public var layers: Int
    public var hiddenSize: Int
    public var queryHeads: Int
    public var kvHeads: Int
    public var maxContextTokens: Int
    public var tokenizer: TokenizerInfo
}

public struct TokenizerInfo: Codable, Equatable, Sendable {
    public var source: String
    public var preserved: Bool
    public var vocabularyPreserved: Bool
    public var chatTemplate: String
}

public struct DeviceProfileConfiguration: Codable, Equatable, Sendable {
    public var sip: String
    public var neuralEngineCores: Int
    public var defaultContextVariant: Int
    public var maxNewTokens: Int
}

public struct AssetInfo: Codable, Equatable, Sendable {
    public var storage: String
    public var prefillPath: String
    public var decodePath: String
    public var tokenizerPath: String?
    public var sha256: String
    public var prefillSHA256: String?
    public var decodeSHA256: String?
    public var tokenizerSHA256: String?
    public var variants: [String: ModelArtifactVariant]?
}

public struct ModelArtifactVariant: Codable, Equatable, Sendable {
    public var deviceProfile: String
    public var prefillPath: String
    public var decodePath: String
    public var tokenizerPath: String?
    public var sha256: String
    public var prefillSHA256: String?
    public var decodeSHA256: String?
    public var tokenizerSHA256: String?
}

public struct SelectedModelArtifact: Codable, Equatable, Sendable {
    public var contextVariant: Int
    public var deviceProfile: String
    public var prefillPath: String
    public var decodePath: String
    public var tokenizerPath: String?
    public var sha256: String
    public var prefillSHA256: String?
    public var decodeSHA256: String?
    public var tokenizerSHA256: String?

    public init(
        contextVariant: Int,
        deviceProfile: String,
        prefillPath: String,
        decodePath: String,
        tokenizerPath: String?,
        sha256: String,
        prefillSHA256: String?,
        decodeSHA256: String?,
        tokenizerSHA256: String?
    ) {
        self.contextVariant = contextVariant
        self.deviceProfile = deviceProfile
        self.prefillPath = prefillPath
        self.decodePath = decodePath
        self.tokenizerPath = tokenizerPath
        self.sha256 = sha256
        self.prefillSHA256 = prefillSHA256
        self.decodeSHA256 = decodeSHA256
        self.tokenizerSHA256 = tokenizerSHA256
    }
}

public struct QuantizationInfo: Codable, Equatable, Sendable {
    public var strategy: String
    public var weights: QuantizationWeights
    public var kvCache: String
    public var structuralReduction: Bool
}

public struct QuantizationWeights: Codable, Equatable, Sendable {
    public var embedding: String
    public var lmHead: String
    public var norms: String
    public var attentionQKO: String
    public var attentionV: String
    public var ffn: String
}

public struct FallbackPolicy: Codable, Equatable, Sendable {
    public var requiresBenchmarkEvidence: Bool
    public var order: [String]
}

enum ModelManifestContract {
    static let expectedModelId = "openbmb/MiniCPM5-1B"
    static let expectedRuntime = "coreml-mlprogram"
    static let explicitKVGraphInterface = "logits-layered-kv"
    static let statefulKVGraphInterface = "stateful-kv"
    static let statefulStepKVGraphInterface = "stateful-step-kv"
    static let supportedGraphInterfaces = [
        explicitKVGraphInterface,
        statefulKVGraphInterface,
        statefulStepKVGraphInterface
    ]
    static let statefulGraphInterfaces = [
        statefulKVGraphInterface,
        statefulStepKVGraphInterface
    ]
    static let supportedKVCacheModes = ["stateful-preferred", "slot-ring", "contiguous-sliding"]
    static let layers = 24
    static let hiddenSize = 1536
    static let queryHeads = 16
    static let kvHeads = 2
    static let headDimension = 128
    static let supportedContextVariants = [256, 512, 1024]
    static let prefillInputIDs = "input_ids"
    static let prefillPositionIDs = "position_ids"
    static let causalMask = "causal_mask"
    static let logits = "logits"
    static let prefillKeyPrefix = "present_key_"
    static let prefillValuePrefix = "present_value_"
    static let decodeTokenID = "token_id"
    static let decodePositionID = "position_id"
    static let decodePastKeyPrefix = "past_key_"
    static let decodePastValuePrefix = "past_value_"
    static let decodeNewKeyPrefix = "new_key_"
    static let decodeNewValuePrefix = "new_value_"

    static func expectedDecodeTokenID(for graphInterface: String) -> String {
        graphInterface == statefulStepKVGraphInterface ? prefillInputIDs : decodeTokenID
    }

    static func expectedDecodePositionID(for graphInterface: String) -> String {
        graphInterface == statefulStepKVGraphInterface ? prefillPositionIDs : decodePositionID
    }
}

File: tools/validation/modelManifest.js
export const SUPPORTED_CONTEXT_VARIANTS = Object.freeze([256, 512, 1024]);
export const SUPPORTED_DEVICE_PROFILES = Object.freeze(["watch-se-2", "watch-se-3"]);
export const SUPPORTED_KV_CACHE_MODES = Object.freeze([
  "stateful-preferred",
  "slot-ring",
  "contiguous-sliding"
]);
export const SUPPORTED_GRAPH_INTERFACES = Object.freeze([
  "logits-layered-kv",
  "stateful-kv",
  "stateful-step-kv"
]);
export const STATEFUL_GRAPH_INTERFACES = Object.freeze([
  "stateful-kv",
  "stateful-step-kv"
]);
export const EXPECTED_MODEL_ID = "openbmb/MiniCPM5-1B";
export const EXPECTED_RUNTIME = "coreml-mlprogram";
export const EXPECTED_GRAPH_SCHEMA = Object.freeze({
  interface: "logits-layered-kv",
  layerCount: 24,
  kvHeads: 2,
  headDimension: 128,
  prefill: Object.freeze({
    inputIDs: "input_ids",
    positionIDs: "position_ids",
    causalMask: "causal_mask",
    logits: "logits",
    keyPrefix: "present_key_",
    valuePrefix: "present_value_"
  }),
  decode: Object.freeze({
    tokenID: "token_id",
    positionID: "position_id",
    causalMask: "causal_mask",
    logits: "logits",
    pastKeyPrefix: "past_key_",
    pastValuePrefix: "past_value_",
    newKeyPrefix: "new_key_",
    newValuePrefix: "new_value_"
  })
});
export const EXPECTED_ARCHITECTURE = Object.freeze({
  layers: 24,
  hiddenSize: 1536,
  queryHeads: 16,
  kvHeads: 2,
  tokenizerSource: EXPECTED_MODEL_ID
});
const SUPPORTED_PRECISIONS = Object.freeze(["fp16", "int8", "int4"]);
const QUANTIZED_WEIGHT_COMPONENTS = Object.freeze([
  "embedding",
  "lmHead",
  "norms",
  "attentionQKO",
  "attentionV",
  "ffn"
]);
const TRANSFORMER_WEIGHT_COMPONENTS = Object.freeze(["attentionQKO", "attentionV", "ffn"]);

export function validateModelManifest(manifest) {
  const errors = [];
  const warnings = [];

  if (!isRecord(manifest)) {
    return {
      ok: false,
      errors: ["manifest must be an object"],
      warnings
    };
  }

  validateModel(manifest, errors);
  validateRuntime(manifest, errors);
  validateArchitecture(manifest, errors);
  validateDeviceProfiles(manifest, errors);
  validateContextVariants(manifest, errors);
  validateAsset(manifest, errors);
  validateQuantization(manifest, errors);
  validateFallbackPolicy(manifest, warnings);

  return {
    ok: errors.length === 0,
    errors,
    warnings
  };
}

export function assertValidModelManifest(manifest) {
  const result = validateModelManifest(manifest);
  if (!result.ok) {
    throw new Error(`Invalid model manifest:\n- ${result.errors.join("\n- ")}`);
  }
}

export function selectContextVariant(manifest, deviceProfile, requestedTokens) {
  assertValidModelManifest(manifest);

  const profile = manifest.deviceProfiles[deviceProfile];
  if (!profile) {
    throw new Error(`Unsupported device profile: ${deviceProfile}`);
  }

  const variants = [...manifest.contextVariants].sort((left, right) => left - right);
  const requested = Number.isFinite(requestedTokens)
    ? requestedTokens
    : profile.defaultContextVariant;
  const selected = variants.filter((variant) => variant <= requested).at(-1);

  return selected ?? variants[0];
}

export function selectModelArtifact(manifest, deviceProfile, requestedTokens) {
  const contextVariant = selectContextVariant(manifest, deviceProfile, requestedTokens);
  const variant = manifest.asset?.variants?.[String(contextVariant)];

  if (variant) {
    return {
      contextVariant,
      deviceProfile: variant.deviceProfile,
      prefillPath: variant.prefillPath,
      decodePath: variant.decodePath,
      tokenizerPath: variant.tokenizerPath ?? manifest.asset.tokenizerPath,
      sha256: variant.sha256,
      prefillSHA256: variant.prefillSHA256,
      decodeSHA256: variant.decodeSHA256,
      tokenizerSHA256: variant.tokenizerSHA256 ?? manifest.asset.tokenizerSHA256
    };
  }

  return {
    contextVariant,
    deviceProfile,
    prefillPath: manifest.asset.prefillPath,
    decodePath: manifest.asset.decodePath,
    tokenizerPath: manifest.asset.tokenizerPath,
    sha256: manifest.asset.sha256,
    prefillSHA256: manifest.asset.prefillSHA256,
    decodeSHA256: manifest.asset.decodeSHA256,
    tokenizerSHA256: manifest.asset.tokenizerSHA256
  };
}

export function summarizeModelManifest(manifest) {
  assertValidModelManifest(manifest);

  return {
    modelId: manifest.model.id,
    runtime: manifest.runtime.type,
    deviceProfiles: SUPPORTED_DEVICE_PROFILES.filter(
      (profile) => profile in manifest.deviceProfiles
    ),
    contextVariants: [...manifest.contextVariants],
    assetStorage: manifest.asset.storage,
    assetVariants: Object.keys(manifest.asset.variants ?? {}).map(Number).sort((left, right) => left - right),
    kvCacheMode: manifest.runtime.kvCacheMode,
    quantizationStrategy: manifest.quantization.strategy,
    kvCachePrecision: manifest.quantization.kvCache
  };
}

function validateModel(manifest, errors) {
  if (manifest.model?.id !== EXPECTED_MODEL_ID) {
    errors.push(`model.id must be ${EXPECTED_MODEL_ID}`);
  }
}

function validateRuntime(manifest, errors) {
  if (manifest.runtime?.type !== EXPECTED_RUNTIME) {
    errors.push(`runtime.type must be ${EXPECTED_RUNTIME}`);
  }

  const entrypoints = manifest.runtime?.entrypoints;
  if (!Array.isArray(entrypoints) || !entrypoints.includes("prefill") || !entrypoints.includes("decode")) {
    errors.push("runtime.entrypoints must include prefill and decode");
  }

  if (!SUPPORTED_KV_CACHE_MODES.includes(manifest.runtime?.kvCacheMode)) {
    errors.push("runtime.kvCacheMode must be stateful-preferred, slot-ring, or contiguous-sliding");
  }

  validateRuntimeGraphSchema(manifest.runtime?.graphSchema, errors);
}

function validateRuntimeGraphSchema(graphSchema, errors) {
  if (!isRecord(graphSchema)) {
    errors.push("runtime.graphSchema must describe Core ML prefill/decode IO");
    return;
  }

  for (const [field, expected] of Object.entries({
    layerCount: EXPECTED_GRAPH_SCHEMA.layerCount,
    kvHeads: EXPECTED_GRAPH_SCHEMA.kvHeads,
    headDimension: EXPECTED_GRAPH_SCHEMA.headDimension
  })) {
    if (graphSchema[field] !== expected) {
      errors.push(`runtime.graphSchema.${field} must be ${expected}`);
    }
  }
  if (!SUPPORTED_GRAPH_INTERFACES.includes(graphSchema.interface)) {
    errors.push("runtime.graphSchema.interface must be logits-layered-kv, stateful-kv, or stateful-step-kv");
  }

  validateNamedSchema("runtime.graphSchema.prefill", graphSchema.prefill, EXPECTED_GRAPH_SCHEMA.prefill, errors);
  validateNamedSchema("runtime.graphSchema.decode", graphSchema.decode, expectedDecodeSchema(graphSchema.interface), errors);
}

function expectedDecodeSchema(graphInterface) {
  if (graphInterface === "stateful-step-kv") {
    return {
      ...EXPECTED_GRAPH_SCHEMA.decode,
      tokenID: "input_ids",
      positionID: "position_ids"
    };
  }
  return EXPECTED_GRAPH_SCHEMA.decode;
}

function validateNamedSchema(path, schema, expectedSchema, errors) {
  if (!isRecord(schema)) {
    errors.push(`${path} must be an object`);
    return;
  }

  for (const [field, expected] of Object.entries(expectedSchema)) {
    if (schema[field] !== expected) {
      errors.push(`${path}.${field} must be ${expected}`);
    }
  }
}

function validateArchitecture(manifest, errors) {
  const architecture = manifest.architecture ?? {};
  if (architecture.layers !== EXPECTED_ARCHITECTURE.layers) {
    errors.push(`architecture.layers must be ${EXPECTED_ARCHITECTURE.layers}`);
  }

  if (architecture.hiddenSize !== EXPECTED_ARCHITECTURE.hiddenSize) {
    errors.push(`architecture.hiddenSize must be ${EXPECTED_ARCHITECTURE.hiddenSize}`);
  }

  if (architecture.queryHeads !== EXPECTED_ARCHITECTURE.queryHeads) {
    errors.push(`architecture.queryHeads must be ${EXPECTED_ARCHITECTURE.queryHeads}`);
  }

  if (architecture.kvHeads !== EXPECTED_ARCHITECTURE.kvHeads) {
    errors.push(`architecture.kvHeads must be ${EXPECTED_ARCHITECTURE.kvHeads}`);
  }

  if (architecture.tokenizer?.source !== EXPECTED_ARCHITECTURE.tokenizerSource) {
    errors.push(`architecture.tokenizer.source must be ${EXPECTED_ARCHITECTURE.tokenizerSource}`);
  }

  if (architecture.tokenizer?.preserved !== true) {
    errors.push("tokenizer must be preserved");
  }

  if (architecture.tokenizer?.vocabularyPreserved !== true) {
    errors.push("vocabulary must be preserved");
  }
}

function validateDeviceProfiles(manifest, errors) {
  if (!isRecord(manifest.deviceProfiles)) {
    errors.push("deviceProfiles must be an object");
    return;
  }

  for (const profile of SUPPORTED_DEVICE_PROFILES) {
    const config = manifest.deviceProfiles[profile];
    if (!isRecord(config)) {
      errors.push(`deviceProfiles.${profile} must be present`);
      continue;
    }

    if (!SUPPORTED_CONTEXT_VARIANTS.includes(config.defaultContextVariant)) {
      errors.push(`deviceProfiles.${profile}.defaultContextVariant must be supported`);
    }
  }
}

function validateContextVariants(manifest, errors) {
  if (!Array.isArray(manifest.contextVariants) || manifest.contextVariants.length === 0) {
    errors.push("contextVariants must be a non-empty array");
    return;
  }

  for (const variant of manifest.contextVariants) {
    if (!SUPPORTED_CONTEXT_VARIANTS.includes(variant)) {
      errors.push(`unsupported context variant ${variant}`);
    }
  }
}

function validateAsset(manifest, errors) {
  if (manifest.asset?.storage === "app-bundle") {
    errors.push("asset.storage must not be app-bundle");
  }

  if (manifest.asset?.storage !== "application-support") {
    errors.push("asset.storage must be application-support");
  }

  if (typeof manifest.asset?.sha256 !== "string" || manifest.asset.sha256.length !== 64) {
    errors.push("asset.sha256 must be a 64-character hex digest");
  }

  for (const field of ["prefillSHA256", "decodeSHA256", "tokenizerSHA256"]) {
    if (field in (manifest.asset ?? {}) && !isSHA256(manifest.asset[field])) {
      errors.push(`asset.${field} must be a 64-character hex digest`);
    }
  }

  validateStatefulSharedArtifacts(manifest, errors);
  validateAssetVariants(manifest, errors);
}

function validateStatefulSharedArtifacts(manifest, errors) {
  if (!STATEFUL_GRAPH_INTERFACES.includes(manifest.runtime?.graphSchema?.interface)) {
    return;
  }

  if (manifest.asset?.prefillPath !== manifest.asset?.decodePath) {
    errors.push("stateful Core ML graphs must use the same artifact path for prefill and decode");
  }

  const variants = manifest.asset?.variants;
  if (!isRecord(variants)) {
    return;
  }

  for (const [contextVariant, variant] of Object.entries(variants)) {
    if (variant.prefillPath !== variant.decodePath) {
      errors.push(`asset.variants.${contextVariant} must use the same artifact path for prefill and decode for stateful Core ML graphs`);
    }
  }
}

function validateAssetVariants(manifest, errors) {
  const variants = manifest.asset?.variants;
  if (variants === undefined) {
    return;
  }

  if (!isRecord(variants)) {
    errors.push("asset.variants must be an object when present");
    return;
  }

  for (const profile of SUPPORTED_DEVICE_PROFILES) {
    const contextVariant = manifest.deviceProfiles?.[profile]?.defaultContextVariant;
    const variant = variants[String(contextVariant)];
    if (!isRecord(variant)) {
      errors.push(`asset.variants.${contextVariant} must be present for ${profile}`);
      continue;
    }

    if (variant.deviceProfile !== profile) {
      errors.push(`asset.variants.${contextVariant}.deviceProfile must be ${profile}`);
    }

    if (typeof variant.prefillPath !== "string" || !variant.prefillPath.endsWith(".mlpackage")) {
      errors.push(`asset.variants.${contextVariant}.prefillPath must be an mlpackage path`);
    }

    if (typeof variant.decodePath !== "string" || !variant.decodePath.endsWith(".mlpackage")) {
      errors.push(`asset.variants.${contextVariant}.decodePath must be an mlpackage path`);
    }

    if (typeof variant.sha256 !== "string" || !/^[a-f0-9]{64}$/.test(variant.sha256)) {
      errors.push(`asset.variants.${contextVariant}.sha256 must be a 64-character hex digest`);
    }

    if (typeof (variant.tokenizerPath ?? manifest.asset?.tokenizerPath) !== "string") {
      errors.push(`asset.variants.${contextVariant}.tokenizerPath must be present`);
    }

    for (const field of ["prefillSHA256", "decodeSHA256", "tokenizerSHA256"]) {
      const value = variant[field] ?? manifest.asset?.[field];
      if (!isSHA256(value)) {
        errors.push(`asset.variants.${contextVariant}.${field} must be a 64-character hex digest`);
      }
    }
  }
}

function validateQuantization(manifest, errors) {
  const quantization = manifest.quantization ?? {};
  if (quantization.strategy !== "mixed-precision-fidelity-first") {
    errors.push("quantization.strategy must be mixed-precision-fidelity-first");
  }

  if (!["fp16", "int8"].includes(quantization.kvCache)) {
    errors.push("quantization.kvCache must be fp16 or int8");
  }

  if (quantization.structuralReduction !== false) {
    errors.push("structuralReduction must be false");
  }

  if (!isRecord(quantization.weights)) {
    errors.push("quantization.weights must describe per-component precision");
    return;
  }

  for (const component of QUANTIZED_WEIGHT_COMPONENTS) {
    if (typeof quantization.weights[component] !== "string") {
      errors.push(`quantization.weights.${component} must be present`);
    } else if (!SUPPORTED_PRECISIONS.includes(quantization.weights[component])) {
      errors.push(`quantization.weights.${component} must be fp16, int8, or int4`);
    }
  }

  validateLayerOverrides(quantization, manifest.architecture?.layers, errors);
}

function validateLayerOverrides(quantization, layerCount, errors) {
  if (quantization.layerOverrides === undefined) {
    return;
  }
  if (!isRecord(quantization.layerOverrides)) {
    errors.push("quantization.layerOverrides must be an object");
    return;
  }

  const resolvedLayerCount = Number.isInteger(layerCount) ? layerCount : EXPECTED_ARCHITECTURE.layers;
  for (const [component, overrides] of Object.entries(quantization.layerOverrides)) {
    if (!TRANSFORMER_WEIGHT_COMPONENTS.includes(component)) {
      errors.push(`quantization.layerOverrides.${component} is not supported`);
      continue;
    }
    if (!isRecord(overrides)) {
      errors.push(`quantization.layerOverrides.${component} must be an object`);
      continue;
    }

    for (const [rawLayer, precision] of Object.entries(overrides)) {
      const layer = Number(rawLayer);
      if (!Number.isInteger(layer)) {
        errors.push(`quantization.layerOverrides.${component} layer must be an integer`);
        continue;
      }
      if (layer < 0 || layer >= resolvedLayerCount) {
        errors.push(`quantization.layerOverrides.${component}.${layer} is outside layer count`);
      }
      if (!SUPPORTED_PRECISIONS.includes(precision)) {
        errors.push(`quantization.layerOverrides.${component}.${layer} must be fp16, int8, or int4`);
      }
    }
  }
}

function validateFallbackPolicy(manifest, warnings) {
  if (manifest.fallbackPolicy?.requiresBenchmarkEvidence !== true) {
    warnings.push("fallbackPolicy.requiresBenchmarkEvidence should be true");
  }
}

function isRecord(value) {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

function isSHA256(value) {
  return typeof value === "string" && /^[a-f0-9]{64}$/.test(value);
}
